full-stack-engineer/submissions/michaelyao/server/src/products/reviews.service.ts

```diff
@@ -0,0 +1,97 @@
+import {
+  BadRequestException,
+  ForbiddenException,
+  Injectable,
+  NotFoundException,
+} from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { CreateReviewDto } from './dto/create-review.dto';
+import { UpdateReviewDto } from './dto/update-review.dto';
+import { Repository } from 'typeorm/repository/Repository';
+import { Product } from '../products/entities/product.entity';
+import { Review } from './entities/review.entity';
+import { User } from '../users/entities/user.entity';
+
+@Injectable()
+export class ReviewsService {
+  constructor(
+    @InjectRepository(Product) private productRepo: Repository<Product>,
+    @InjectRepository(Review) private reviewRepo: Repository<Review>,
+  ) {}
+
+  async create(
+    user: User,
+    product_id: number,
+    createReviewDto: CreateReviewDto,
+  ) {
+    const [product] = await this.productRepo.find({
+      where: {
+        id: product_id,
+      },
+      relations: ['reviews'],
+    });
+    if (!product) {
+      throw new NotFoundException('product not found');
+    }
+    const reviewUserIds = product.reviews.map((review) => review.user.id);
+    if (reviewUserIds.includes(user.id)) {
+      throw new BadRequestException(
+        'you have already created an review for the product',
+      );
+    }
+
+    const review = this.reviewRepo.create(createReviewDto);
+    review.product = product;
+    review.user = user;
+    return this.reviewRepo.save(review);
+  }
+
+  async findAll(product_id: number) {
+    const reviews = await this.reviewRepo.find({
```

Comment:
Which order will be used to return the reviews? I guess by default is `createdAt` ascending. But normally will be the newest first.

Reply:
Yep. The default is `ascending`. I asked the boss. I was told that the latest should stay in the bottom. So I guess `ascending` works for the project.
<img width="405" alt="image" src="https://user-images.githubusercontent.com/22899096/158300493-42e4ac9f-6f05-47e2-a34b-dab2fa12d8ee.png">